ops/testing.py

```diff
@@ -319,16 +319,33 @@ def set_leader(self, is_leader=True):
         if is_leader and not was_leader and self._charm is not None and self._hooks_enabled:
             self._charm.on.leader_elected.emit()
 
+    def get_backend_calls(self, reset=False):
+        """Return the calls that we have made to the TestingModelBackend.
+
+        This is useful mostly for testing the framework itself, so that we can assert that we
+        do/don't trigger extra calls.
+
+        :param reset: If True, reset the calls list back to empty.
+        :type reset: bool
+        :return: [(call1, args...), (call2, args...)]
+        """
+        calls = self._backend._calls.copy()
```

Comment:
We shouldn't use private attributes from other objects.

Or the backend should allow us to use the attribute (make it public), or provide a way to access those calls.

Don't be shy on making `calls` a public attribute, it's in a private class anyway so it shouldn't be used from outside the harness anyway...

Reply:
So, the model that I used was that _TestModelBackend's public API is strictly conforming to what Model needs from ModelBackend, and than Harness was the place that kept the consistency of the _TestModelBackend's internal structures.
We *can* add a bunch of public methods on _TestModelBackend, but that confuses what exists to be a ModelBackend, and what exists to be a backend-that-can-be-mutated-by-Harness.

If we want to change that pattern, I'm open to the discussion. But I really did want _TestModelBackend to be mostly dumb, and Harness to be the glue that kept it consistent. (put another way, the consistency is in Harness, and _TestModelBackend is just an implementation detail of *Harness* that is passed to Model.)

Reply:
I should also note that we had some discussions about accessing private methods from Harness of other classes. Namely Harness also does poke at private internals of Model (which is a much bigger no-no than private internals of the coupled class in the same module).
However, the discussion between Gustavo and I ended up with "Harness is the official, supported way for 3rd parties to poke at internals of Model, rather than exposing those knobs as 'public' on Model.". (namely _invalidate() on a couple of the cache objects could be made public.)
So while it overly ties Harness to the other Framework classes, since that is the purpose of Harness, we can move forward with it.
It is true that some aspects of Harness are going to be more brittle than they might be otherwise, but it also avoids having to publish "public for others" attributes for things that we want "private to the framework, but supported between framework objects.".

Reply:
Perfect. May you please improve the _TestModelBackend docstring with some of this information? Specially stating that is responsibility of the Harness class to ensure consistency between `_relation_ids_map`, `_relation_names`, `_relation_list_map`, and `_relation_data`...

Thanks!!

Reply:
added.